README
What is Cexp?
-------------

 -- Brought to you by Till Straumann <strauman@slac.stanford.edu> --
 -- $Id: README,v 1.25 2009/03/12 17:34:22 strauman Exp $ --

The Cexp utility is a C-expression interpreter which
gives its user access to all symbols present in the currently
executing program. Its primary target is RTEMS, an open source
RTOS, but Cexp should compile and run on virtually any platform
supported by the BFD library.

If Cexp is linked to an executable, it is
possible to invoke arbitrary functions and to
read/write variables (and hence virtually any
memory location) by interpreting C-style expressions.

Cexp can access object files using either the pmbfd/pmelf
libraries ('poor-man's BFD) that comes with Cexp or libbfd.
The latter is more powerful:
 - not limited to the ELF object format.
 - supports many more machine types (pmbfd currently 
   only supports 32-bit powerpc, m68k and i386 and 64-bit
   x86_64).
 - using BFD / libopcodes, a disassembler comes for free.

Otoh, using pmbfd/pmelf has the following advantages:
 - much smaller memory footprint.
 - less restrictive license. Linking against the BFD
   library is subject to the GPL. See LICENSING
   below for more information.

What is New?
------------

 - for later versions, consult the ChangeLog file.

 - 2003/02/14: CEXP_1.2.beta ("Valentine") released.
   It supports G++ prioritized execution of static
   constructors/destructors.

 - 2003/02/13: CEXP_1.1.beta is released. It fixes a bug
   of the exception handler registration.
   
   Makefile.syms has been added and modularized copies
   of the RTEMS 'cdtest' sample program.

Description
-----------

CEXP is a simple utility featuring
 - symbol table
 - C expression parser/interpreter
 - type "engine"
 - user definable variables
 - recursive invocation/simple scripts
 - runtime 'module' loader/linker
 - disassembler (only if linked with BFD)
 - CEXP is reentrant

Cexp knows about the basic C types (no aggregates), i.e.

char,   char*
short,  short*
long,   long*
float,  float*
double, double*
long    (*)()
double  (*)()

and interprets C-style expressions, e.g.

  Cexp> printf("Hello world")

  Cexp> some_variable = 0xdeadbeef

  Cexp> some_double_variable = *(double*)&some_variable

  Cexp> a=printf, a && a("The square root of 2 is %g\n",sqrt(2.0))

Symbol Table
------------
A) Using a Symbol Table File
- - - - - - - - - - - - - - -
On startup, Cexp reads the 'system' symbol table either from 
the executable itself, or from a stripped-down '.sym'
file - such a stripped-down symbol file can be created using
the 'objcopy' (preferred, but needs binutils 2.18 or later)
or 'xsyms' (deprecated; comes with Cexp) utilities.

B) Builtin Symbol Table File
- - - - - - - - - - - - - - -
The 'xsyms' tool can be used to generate
a symbol table in C-source form for subsequent compilation and
linkage into the final executable. This usually involves linking
the executable twice:
  1) compile and link all sources, libraries etc.; build an executable
  2) xsyms -C executable  mysymtab.c
  3) compile mysymtab.c (set -I to Cexp installation 'include' directory
     or Cexp source directory, since <cexpsyms.h> is included)
  4) link application again, but this time add 'mysymtab.c' to the
     list of sources.
  NOTES: Step 1) linking succeeds without a symtab.o because there's
         a weak NULL-ptr alias for the builtin symtab.
		 When building the demo program you can use a builtin symbol
		 table:
		   a) build demo the normal way (corresponds to step 1 above)
		   b) type 'make builtin-symtab'
		      this extracts symbol table info from the executable
			  created in step a) and creates an object file
			  'cexp-builtin-symtab.o' (steps 2+3 above).
		   c) type 'make' again. This re-links the demo program
		      now using the 'cexp-builtin-symtab.o' object
			  (corresponds to step 4 above).
A similar source file can also be generated using the 'ldep' utility
(discussed elsewhere, it is part of RTEMS-GeSys).

Without access to a symbol table file which also provides Cexp
with information about the architecture it is executing on, Cexp
might fail to detect the correct BFD CPU architecture which is
needed for the disassembler. In such cases, the '-a' command
line option can be used to specify the architecture (arch string
can be obtained by running 'objdump -f' on the ELF executable).

Symbol Type
- - - - - -

Once the symbol table is read, Cexp guesses the symbol/variable
types from the symbol sizes. (Guess is only supported on ELF targets.)

In some cases, this guess is obviously wrong, but you can 
always cast the symbol to the correct type.

Arrays are normally 'void' like every symbol for which's
type Cexp cannot make a guess. It is still possible, however, to
use such a symbol, simply by taking its address and
casting to a different pointer type:
    
Cexp> *(((long*)&some_array)+25)

evaluates to the unsigned long array element
some_array[25]. Alternatively, you can redeclare
its type:

Cexp> long *some_array !
Cexp> *(some_array + 25)

Note the exclamation mark: terminating a variable declaration with an
exclamation mark allows you to redefine its type. Only new variables
may be declared without the '!'. This protects you from erroneously
use existing symbols.

Cexp provides a series of lookup functions
for symbols. Most noteworthy:

lkup(char *regexp)
lkaddr(unsigned long addr)

These are just normal C functions which can be
invoked by Cexp. Note that lkup() takes a regular
expression argument allowing for powerful searching.

C-Expression Parser / Interpreter
---------------------------------
The C expression parser has some restrictions:
  - no '?' ':' expression
  - no [], '->'
  - the '.' operator has a special meaning - see below
  - no multilevel pointer dereferencing (since e.g. char**
    is not a valid type, Cexp cannot dereference 
    **addr. You must explicitly cast this:

      Cexp> *(char*)*(long*)some_char_pointer_pointer

which assumes that sizeof(long) == sizeof(char*)
for the particular machine.

However, most of the other operators are available, including
the 'comma' operator and the logical '&&' and '||' with the
correct semantics, i.e. conditional evaluation of expression
parts e.g.

  Cexp> (dd=deviceDetect(_some_device_address)) && driverInit(dd)

will call the driverInit() function only if deviceDetect()
returns a nonzero value.

The '.' operator (structure field access) has a special
meaning to Cexp: symbol "member" (in an OOP sense) access.
Currently, the only member function defined is 'help'. Consult
the 'Help' section about details.

Function calling:
The user is responsible for feeding properly typed arguments;
unused arguments will be filled with integral/scalar 0, which on 
most ABIs is safe.
Since the symbol table (on ELF, Cexp is not using .stab but .symtab)
provides no info about a function's return type, Cexp assumes
all functions to return long. Floating point functions
must be cast appropriately:

  Cexp> ((double(*)())sqrt)(2.345)

Or

  Cexp> double (*sqrt)()
  Cexp> sqrt(2.345)

A lot of type casting can be avoided by re-declaring or
using user defined variables, see below.


Type Engine
-----------
Cexp knows about (and only about) the primitive types listed
above.
NOTE: Cexp treats ALL integral types as UNSIGNED although
no 'unsigned' keyword is used or recognized. The main
reason for this is to save typing. Hence
char is in fact unsigned char etc.

An 'int' type is missing (but could easily added) which
means that on some machines (e.g. 64-bit cpus) no 32bit
type is currently available...

User Variables
--------------
Cexp supports (fully typed) user variables which are
visible in a global namespace (i.e. visible to other
instances of Cexp). The name of user variables must
not collide with symbols present in the symbol table.
(When loading object files [AKA 'modules'], name clashes
are ignored. A symbol in a newly loaded module that
conflicts with an already defined user variable silently
takes precedence [object file symbols have a higher
priority than user variables]. However, after unloading
the conflicting module, the user variable becomes again
'visible'.)

A user variable is created simply by assigning it
a value which also automatically defines its type.

    hallo="hallo"

creates a 'char*' and assigns it the address of a
string constant. The string is stored in 'malloc()ated'
memory and 'lives' forever. Subsequent use of the
same string (strcmp(a,b)==0) results in re-using the
already stored instance of the string.
String constants must not be written-to!

The type of user variables can be modifed, simply
by re-declaring them, in a C-style manner:

    long hallo

results in 'hallo' maintaining its value (the address
of the string constant) but interpreting it as a long.
Hence

    chppt = &hallo

is an 'disguised' char** and

    *(char*)*chppt

yields 'h'

User variables can also hold function pointers and hence
can be handy abbreviations for long names and casts.
E.g. a convenient variable can be set for 'sqrt()'
(which returns a double value):

    s=((double(*)())sqrt)

It is then possible to automatically get the correct
return type:

    printf("Let's print a double: %g\n",s(44.34))

The special user variable 'ans' stores the result
of the last successfully evaluated expression. E.g:

    Cexp> 3+4
	0x00000007 (7)
	Cexp> ++ans
	0x00000008 (8)
	Cexp> printf("%u\n",ans)
	8
	0x00000002 (2)
	Cexp>

The 'ans' variable is special in the sense that unlike
other user variables it is not global but local to the
executing parser, i.e., parsers running in different
task contexts each maintain their own instances of
'ans'.

Help Facility
-------------

The '.' operator (structure field access) has a special
meaning to Cexp: symbol "member" (in an OOP sense) access.
Currently, the only member function defined is 'help'. I.e.
by typing

Cexp>  some_symbol.help()

You get help information about that symbol. If no help
is defined, or if you request 'verbose' help:

Cexp> some_symbol.help(1)

the symbol address, value and type information are printed
(same format as the 'lkup' output).

You can add/modify help information to a symbol (e.g. a user
defined variable) simply by providing a string argument to
the symbol's help member:

Cexp> long myvar
Cexp> myvar.help("I just created a 'long' variable")

Note that help is stored 'per symbol', i.e. when creating
an 'alias' as in the example above:

Cexp> s=(double (*)())sqrt

the help information is not propagated to 's'. If 'sqrt'
had help information, it would have to be copied:

Cexp> s.help(sqrt.help())

This example shows that the return value of the 'help() member'
is the address of the (static) help text string.

How to Start/Invoke Cexp - Simple Scripts
-----------------------------------------
Cexp has two entry points,

    cexp_main(int argc, char **argv)

and

    cexpsh(char *arg1,...)

i.e. a 'main' style and a 'vararg' style. The
'vararg' version ends up building an argument list
and calling cexp_main(). It is mainly intended
for recursively invoking 'cexpsh()' e.g. for 
reading a series of lines from a file.
Note that the string argument list submitted to
cexpsh() must be NULL terminated. When calling
cexpsh() from cexpsh() (e.g. to interpret a script),
this is not necessary, however, because any unused
arguments (up to the internal maximum of 10 args)
are filled with zeroes...

The calling syntax of cexp_main/cexpsh is as follows:

cexpsh [-s <symbolfile>] [-h] [-d] [-q] [<scriptfile>]

-h and -d have the obvious effect of printing usage
info and enabling debugging information (only available
if configured with --enable-YYDEBUG).

The -q ('quiet') flag instructs Cexp not to print
any normal output (errors are still reported to stderr)
on stdout. This behavior can be desirable when evaluating
scripts.

Note that (unless you are using a built-in symbol
table) the -s option MUST be used the first time
'Cexp' is started and it must be provided with an
appropriate symbol file. This can be the executable
itself of a stripped version (use the 'xsyms' utility)
to reduce memory usage and loading time on RTEMS systems.
A basic check is made to protect against version mismatch
between the symbol file and the executable.

Once Cexp has loaded the system's symbol table, further
instances will simply use the global system table if
the -s argument is missing.

Cexp then reads commands from stdin (using the TECLA
library) or alternatively from a script file.

Cexp ignores any characters present on a line
after it scans the comment tokens
'#' or '//'.

Example for invoking Cexp from Cexp to evaluate
a script:

  Cexp> cexpsh("script_file")

While the former syntax involves recursive execution of
Cexp, it is now also possible to let the current instance
divert its input to read from a script using the
'<' operator:

  Cexp> < script_file

There are slight syntactic differences when using this syntax,
though. Evaluation of the file name is not performed by
the parser but by a simple preprocessor. The filename doesn't have
to be embedded in double quotes (but it can be). Note that
no escape sequences are recognized. It is possible, however,
to use filenames with embedded whitespace by enclosing the
script name in single or double quotes. In this case, the
entire string between matching quotes is used. E.g., (RHS
string delimiter is '"' -- remember that there are no escape
sequences...):

  < a b c     --> read from file "a"
  < 'a b' "c" --> read from file "a b"
  < "a'b"     --> read from file "a'b"
  < "a\"b"    --> read from file "a\\" (a followed by a single backslash)


Loadable Modules / Runtime Loader
---------------------------------

When built aginst the pmbfd or BFD library, Cexp is capable of
dynamically loading object files into a running program.
Cexp keeps track of module dependencies. Note that this
only covers symbol table dependencies. Cexp rejects
unloading a module 'A' if there is still another module,
'B' loaded which had undefined symbols resolved against 'A's
symbols. Obviously, more subtle dependencies, such as
threads using a modules text or data cannot be tracked
easily and are ignored.

An object file (AKA 'module') can be loaded invoking the
command:

  Cexp> someModule=cexpModuleLoad("someModule.o")

The loader returns a 'module ID', which in this example
is stored in the user variable 'someModule'. If errors
occur during the load (such as undefined references or
multiple symbol definitions), they are reported and a NULL
module ID is returned.

After loading, C++ static constructors are executed and
exception handler frames are registered. Note that these
features are probably only supported on ELF and, especially
the exception handling, might only work for gcc compiled
code. Unfortunately, even different versions of gcc and/or
target architectures involve varying implementations of
exception handling - YMMV...

An (unused) module can be unloaded by passing its ID to
cexpModuleUnload() (prior to unloading, the C++ static
destructors are executed):

  Cexp> cexpModuleUnload(someModule)

Two more routines are useful in this context:

  cexpModuleInfo([ID])

prints info about a specific module (ID) or all currently
loaded modules if passed NULL ID.

  cexpModuleFindByName("regexp")

searches the list of modules for a regular expression and
reports the IDs of all matches. It returns the first ID
found (or NULL if there was no match).

Search Path for Loadable Modules
- - - - - - - - - - - - - - - - -
If set, the PATH environment variable is consulted
by the loader when it tries to locate object files
by subsequently prepending colon-separated search
paths as listed in PATH to the file name. The search
starts with the first component and stops as soon
as the file is found. An empty search path (two colons
in a row or a leading or trailing colon) is equivalent
to the current directory ('.'). Note that the current
directory is not searched if PATH is set but does not
contain the CWD. PATH is ignored if the object name
contains a (relative or absolute) directory path already.

Example: search '/tmp', '/TFTP/BOOTP_HOST/blah/bin' and
finally the CWD:

setenv("PATH","/tmp:/TFTP/BOOTP_HOST/blah/bin:",1)

Building loadable modules
- - - - - - - - - - - - -
Note the important difference between a run-time loaded/linked
object and a _shared_ object (such as a shared library). The 
latter, although also 'run-time loaded' is different from
the former. A shared object is shared among several entities
using disjunct address spaces (e.g. different UNIX processes).
Supporting shared objects involves PIC and GOTs.

***********************************************************

CEXP does _not_ support shared libraries! An attempt to run
a loaded module that was compiled with '-fpic', '-fPIC'
and/or linked with the '-shared' options will crash hard.

***********************************************************


To illustrate the difference, consider a LINUX system
running two instances of the 'cexp' demo. 'cexp' is linked
against glibc which is a shared library - both instances
of 'cexp' use the _same_ copy of libc residing in physical memory.
Let's now assume the both of the demo programs issue

	cexpModuleLoad("someObjectFile.o")

Cexp's run-time loader doesn't support shared objects - hence
both instances of 'cexp' will end up with their own copy of
'someObjectFile' which will get loaded twice to physical memory.

The main target of Cexp is RTEMS, a real-time OS which has
a global address space shared by all threads. Hence, there is
no need for shared object support but run-time loading
code is still very desirable.

The Cexp run-time loader simply accepts _any_ relocatable object file.
The simplest modules are just object files:

	cross-gcc -c -O some_object.c

Multiple objects can be combined using the '-r' linker option

	cross-ld  -o some_object.o -r some1.o some2.o some3.o

It is also possible to convert entire (static) libraries into
loadable objects:

	cross-ld  -o lib_object.o -r --whole-archive libSome.a

The 'system' symbol table is itself an object file from where
Cexp loads the initial symbol table at startup. The executable
itself may be used for that purpose (unless it is a pure binary
as it is the case on some embedded systems). Alternatively
(for saving memory and time), a stripped-down object file containing
only the symbol table may be generated using the 'objcopy'
tool, e.g:

	cross-objcopy --extract-symbol cexp cexp.sym
	cexp -s cexp.sym

This requires binutils 2.18 or later. Alternatively, the
'xsyms' utility that comes with CEXP may be used

	xsyms cexp cexp.sym

but this is deprecated.

Weak Symbol Support
- - - - - - - - - -
Cexp can handle weak symbol references (including weak
undefined symbols). HOWEVER, this works only one-way:
If a module with a strong symbol is already loaded
(or if the base system exports a strong definition)
then loading a module with a weak reference works as
expected, i.e., the references are bound to the existing
strong definition.

OTOH, if a weak definition exists (either in the system
symbol table or a previosly loaded module) then it is
NOT possible to load another module with a strong
definition. (If you think about it: all relocations of
the already linked and loaded stuff would have to be
redone -- including the relocations within the base
executable...).

When a module with a weak undefined reference is loaded
then (provided that no existing definition is found in
the system or other modules) a new NULL symbol is created
which is otherwise treated like a newly defined COMMON
symbol.


Loadable Modules 'Magic'
- - - - - - - - - - - - -
When loading modules, 'Cexp' does some magic operations on
special symbols it recognizes. Note that the all-capital names
given here are macros - the actual names can be found in the
header.

	CEXPMOD_INITIALIZER_SYM (defined in cexpmodP.h), when
		present in a loaded module, a routine with this
		name is invoked by cexpModuleLoad() just after calling
		C++ constructors.
		This 'module-constructor' routine [for C++ information
		see below] can be used to initialize a (non-C++) module.

	CEXPMOD_FINALIZER_SYM, when present in a loaded module,
		a routine with this name is invoked just prior to calling
		C++ destructors and unloading the module.
		The FINALIZER may reject the unloading attempt by
		returning a nonzero value.

	CEXP_HELP_TAB (defined in cexpHelp.h). A module may define

debian/changelog
rtems-cexp (2.0b-8) unstable; urgency=low

  * Minor makefile change
  * rebuild for jessie

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Sun, 10 May 2015 18:22:15 -0400

rtems-cexp (2.0b-7) unstable; urgency=low

  * Change dependency to rtems-common (>=4.9.6-1~)
  * Remove rtems-beatnik to satisfy dependencies

 -- Leonid Flaks <flaks@bnl.gov>  Tue, 29 Oct 2013 13:25:39 -0400

rtems-cexp (2.0b-6) unstable; urgency=low

  * All binary packages are host independent

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Tue, 18 Sep 2012 14:20:51 -0400

rtems-cexp (2.0b-5) unstable; urgency=low

  * auto detect rtems series
  * build for beatnik BSP
  * Build for RTEMS 4.10

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Tue, 18 Sep 2012 13:45:56 -0400

rtems-cexp (2.0b-4) unstable; urgency=low

  * Build for mvme2307

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Wed, 27 Jan 2010 21:22:05 -0500

rtems-cexp (2.0b-3) unstable; urgency=low

  * Generate dependency list using rtems-debhelper

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Fri, 18 Dec 2009 13:48:27 -0500

rtems-cexp (2.0b-2) unstable; urgency=low

  * Add dependency on specific RTEMS version to aid upgrades

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Thu, 17 Dec 2009 16:55:37 -0500

rtems-cexp (2.0b-1) unstable; urgency=low

  * Initial release

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Tue, 21 Jul 2009 14:23:28 -0400



pmbfd/pmbfd.h
/* $Id: pmbfd.h,v 1.17 2009/03/12 04:42:14 strauman Exp $ */

/* 
 * Authorship
 * ----------
 * This software ('pmbfd' BFD emulation for cexpsh) was created by
 *     Till Straumann <strauman@slac.stanford.edu>, 2008,
 * 	   Stanford Linear Accelerator Center, Stanford University.
 * 
 * Acknowledgement of sponsorship
 * ------------------------------
 * This software was produced by
 *     the Stanford Linear Accelerator Center, Stanford University,
 * 	   under Contract DE-AC03-76SFO0515 with the Department of Energy.
 * 
 * Government disclaimer of liability
 * ----------------------------------
 * Neither the United States nor the United States Department of Energy,
 * nor any of their employees, makes any warranty, express or implied, or
 * assumes any legal liability or responsibility for the accuracy,
 * completeness, or usefulness of any data, apparatus, product, or process
 * disclosed, or represents that its use would not infringe privately owned
 * rights.
 * 
 * Stanford disclaimer of liability
 * --------------------------------
 * Stanford University makes no representations or warranties, express or
 * implied, nor assumes any liability for the use of this software.
 * 
 * Stanford disclaimer of copyright
 * --------------------------------
 * Stanford University, owner of the copyright, hereby disclaims its
 * copyright and all other rights in this software.  Hence, anyone may
 * freely use it for any purpose without restriction.  
 * 
 * Maintenance of notices
 * ----------------------
 * In the interest of clarity regarding the origin and status of this
 * SLAC software, this and all the preceding Stanford University notices
 * are to remain affixed to any copy or derivative of this software made
 * or distributed by the recipient and are to be affixed to any copy of
 * software made or distributed by the recipient that contains a copy or
 * derivative of this software.
 * 
 * ------------------ SLAC Software Notices, Set 4 OTT.002a, 2004 FEB 03
 */ 
#ifndef PMBFD_H
#define PMBFD_H

#include <stdio.h>
#include <stdint.h>

#define _PMBFD_

#ifdef __cplusplus
extern "C" {
#endif

typedef struct bfd_arch_info_type bfd_arch_info_type;

typedef unsigned long bfd_vma;
typedef          long bfd_signed_vma;
typedef unsigned long bfd_size_type;
typedef unsigned long symvalue;

/*
 * Here's a trick: 
 * BFD ansidecl.h defines PTR void*
 * and it is sometimes included before bfd.h
 * sometimes after.
 * If we define PTR here then ansidecl.h included
 * after bfd.h would redefine it.
 * By making this a typedef, including ansidecl.h
 * afterwards does no harm. OTOH, if ansidecl.h was
 * included before us then its definition masks
 * our typedef ;-)
 */
#ifndef PTR
typedef void *PTR;
#endif

/*
 * we do need to define TRUE/FALSE namely if
 * opcodes is used together with pmbfd...
 */
#ifndef TRUE
#define TRUE  1
#elif (TRUE) != 1
#warning "TRUE already defined to a value other than 1!"
#endif

#ifndef FALSE
#define FALSE 0
#elif (FALSE) != 0
#error   "FALSE already defined to a value other than 0!"
#endif

typedef long          file_ptr;

typedef unsigned int flagword;
typedef uint8_t      bfd_byte;

struct bfd;

typedef struct bfd bfd; 

/* Symbol attributes */
#define BSF_KEEP			0x0020
#define BSF_SECTION_SYM		0x0100
#define BSF_FUNCTION		0x0010
#define BSF_OLD_COMMON		0x0200
/* NOTE: BSF_OBJECT definition differs from BFD
 *       so that we can store the flags in a
 *       16-bit word.
 *
#define BSF_OBJECT			0x10000
 */
#define BSF_OBJECT			0x1000
#define BSF_GLOBAL			0x0002
#define BSF_LOCAL			0x0001
#define BSF_WEAK			0x0080

/* The following are not neede by cexp but
 * exist for testing against objdump -t
 */
#define BSF_FILE			0x4000
#define BSF_DEBUGGING		0x0008

/* BSF_THREAD_LOCAL is 0x40000 in BFD
 * we only support it so we can strip
 * TLS symbols when creating a symbol
 * table for the cexp demo program.
 */
#define BSF_THREAD_LOCAL	0x2000

#define SEC_ALLOC			0x00000001
#define SEC_RELOC			0x00000004
#define SEC_GROUP			0x04000000
#define SEC_LINK_ONCE		0x00020000
#define SEC_LINK_DUPLICATES	0x00040000

/* The following are not needed by cexp but defined
 * for testing against BFD objdump -h..
 */
#define SEC_LOAD			0x00000002
#define SEC_READONLY		0x00000008
#define SEC_CODE			0x00000010
#define SEC_DATA			0x00000020
#define SEC_ROM				0x00000040
#define SEC_CONSTRUCTOR		0x00000080
#define SEC_HAS_CONTENTS	0x00000100
#define SEC_DEBUGGING		0x00002000
#define SEC_EXCLUDE			0x00008000
#define SEC_LINK_DUPLICATES_DISCARD		0x00000000
#define SEC_LINK_DUPLICATES_ONE_ONLY	0x00080000
#define SEC_LINK_DUPLICATES_SAME_SIZE	0x00100000
#define SEC_LINK_DUPLICATES_SAME_CONTENTS \
	(SEC_LINK_DUPLICATES_ONE_ONLY | SEC_LINK_DUPLICATES_SAME_SIZE)
#define SEC_MERGE			0x01000000
#define SEC_STRINGS			0x02000000

typedef struct sec asection;

typedef struct {
	const char      *name; /* Beware; two asymbols may point to the same name */
	/* NOTE: To make 'asymbol' smaller the 'val' field has slightly
	 *       different semantics than demanded by BFD (BFD stores the
	 *       symbol size of common symbols in 'value').
	 *       Compatible semantics are obtained by accessing the value
	 *       via the 'bfd_asymbol_value() / pmbfd_asymbol_set_value()'
	 *       routines. 
	 *       The member name differs from BFD's so we can detect attempts
	 *       to tamper with 'value' w/o going through the proper access
	 *       macros/routines.
	 */
	symvalue        val;
	bfd_size_type   size;
	uint16_t        flags;
	uint16_t        secndx;
} asymbol;

#define bfd_asymbol_name(s) ((s)->name)

asection *
elf_next_in_group(asection *);

int
elf_get_size(bfd *abfd, asymbol *asym);

unsigned
elf_get_align(bfd *abfd, asymbol *asym);

#define align_power(addr, align)    \
  (((addr) + ((bfd_vma) 1 << (align)) - 1) & ((bfd_vma) -1 << (align)))

extern asection *bfd_abs_section_ptr;
extern asection *bfd_und_section_ptr;
extern asection *bfd_com_section_ptr;

enum bfd_architecture {
	bfd_arch_unknown,
	bfd_arch_obscure,
	bfd_arch_m68k         =  2,
#define bfd_mach_m68000                         1
#define bfd_mach_m68008                         2
#define bfd_mach_m68010                         3
#define bfd_mach_m68020                         4
#define bfd_mach_m68030                         5
#define bfd_mach_m68040                         6
#define bfd_mach_m68060                         7
#define bfd_mach_cpu32                          8
#define bfd_mach_mcf_isa_a_nodiv                9
#define bfd_mach_mcf_isa_a                     10
#define bfd_mach_mcf_isa_a_mac                 11
#define bfd_mach_mcf_isa_a_emac                12
#define bfd_mach_mcf_isa_aplus                 13
#define bfd_mach_mcf_isa_aplus_mac             14
#define bfd_mach_mcf_isa_aplus_emac            15
#define bfd_mach_mcf_isa_b_nousp               16
#define bfd_mach_mcf_isa_b_nousp_mac           17
#define bfd_mach_mcf_isa_b_nousp_emac          18
#define bfd_mach_mcf_isa_b                     19
#define bfd_mach_mcf_isa_b_mac                 20
#define bfd_mach_mcf_isa_b_emac                21
#define bfd_mach_mcf_isa_b_float               22
#define bfd_mach_mcf_isa_b_float_mac           23
#define bfd_mach_mcf_isa_b_float_emac          24
	bfd_arch_i386         =  9,
#define bfd_mach_i386_i386	                    1
#define bfd_mach_i386_i8086                     2
#define bfd_mach_i386_i386_intel_syntax         3
#define bfd_mach_x86_64                        64
#define bfd_mach_x86_64_intel_syntax           65
	bfd_arch_powerpc      = 21,
#define bfd_mach_ppc                           32
#define bfd_mach_ppc64                         64
#define bfd_mach_ppc_403                      403
#define bfd_mach_ppc_403gc                   4030
#define bfd_mach_ppc_505                      505
#define bfd_mach_ppc_601                      601
#define bfd_mach_ppc_602                      602
#define bfd_mach_ppc_603                      603
#define bfd_mach_ppc_ec603e                  6031
#define bfd_mach_ppc_604                      604
#define bfd_mach_ppc_620                      620
#define bfd_mach_ppc_630                      630
#define bfd_mach_ppc_750                      750
#define bfd_mach_ppc_860                      860
#define bfd_mach_ppc_a35                       35
#define bfd_mach_ppc_rs64ii                   642
#define bfd_mach_ppc_rs64iii                  643
#define bfd_mach_ppc_7400                    7400
#define bfd_mach_ppc_e500                     500
	bfd_arch_rs6000        = 22
#define bfd_mach_rs6lk                       6000
#define bfd_mach_rs6lk_rs1                   6001
#define bfd_mach_rs6lk_rsc                   6002
#define bfd_mach_rs6lk_rs2                   6003
};

enum bfd_flavour {
	bfd_target_unknown_flavour = 0,
	bfd_target_elf_flavour = 5
};

enum bfd_endian {
	BFD_ENDIAN_BIG,
	BFD_ENDIAN_LITTLE,
	BFD_ENDIAN_UNKNOWN
};

typedef int bfd_boolean;

typedef enum bfd_format {
	bfd_unknown = 0,
	bfd_object  = 1,
	bfd_archive = 2,
	bfd_core    = 3
} bfd_format;

typedef enum bfd_reloc_status {
	bfd_reloc_ok = 0,
	bfd_reloc_overflow,
	bfd_reloc_outofrange, /* section doesn't contain address */
	bfd_reloc_continue,
	bfd_reloc_notsupported,
	bfd_reloc_other,
	bfd_reloc_undefined,  /* reloc. against undef. symbol    */
	bfd_reloc_dangerous
} bfd_reloc_status_type;

enum bfd_architecture bfd_get_arch(bfd *abfd);

unsigned long bfd_get_mach(bfd *abfd);

int
bfd_big_endian(bfd *abfd);

int
bfd_little_endian(bfd *abfd);

bfd*
bfd_asymbol_bfd(asymbol *sym);

bfd_vma
bfd_asymbol_value(asymbol *sym);

long
bfd_canonicalize_symtab(bfd *abfd, asymbol**);

bfd_boolean
bfd_check_format(bfd *abfd, bfd_format format);

bfd_boolean
bfd_close_all_done(bfd *abfd);

#if 0
bfd_boolean
bfd_discard_group(bfd *, asection *);
#endif

#define HAS_SYMS	0x10

flagword
bfd_get_file_flags(bfd *abfd);

enum bfd_flavour
bfd_get_flavour(bfd *abfd);

long
bfd_get_reloc_upper_bound(bfd *abfd, asection *sect);

asection *
bfd_get_section(asymbol *sym);

/* Note: this actually returns ld(alignment)! */
unsigned int
bfd_get_section_alignment(bfd *abfd, asection *sect);

bfd_size_type
bfd_get_section_size(asection *section);

bfd_boolean
bfd_get_section_contents(bfd *abfd, asection *section, void *location, file_ptr offset, bfd_size_type count);

flagword
bfd_get_section_flags(bfd *abfd, asection *sect);

const char *
bfd_get_section_name(bfd *bfd, asection *sect);

bfd_vma
bfd_get_section_vma(bfd *bfd, asection *sect);

bfd_vma
bfd_get_section_lma(bfd *bfd, asection *sect);

long
bfd_get_symtab_upper_bound(bfd *bfd);

char *
bfd_get_target(bfd *abfd);

char *
bfd_get_unique_section_name(bfd *abfd, const char *template, int *count);

void
bfd_init(void);

int
bfd_is_abs_section(asection *sect);

int
bfd_is_com_section(asection *sect);

int
bfd_is_und_section(asection *sect);

asymbol *
bfd_make_empty_symbol(bfd *bfd);

asection *
bfd_make_section(bfd *, const char *name);

void
bfd_map_over_sections(bfd *abfd, void (*f)(bfd *abfd, asection *sect, void *closure), void *closure);

bfd *
bfd_openstreamr(const char *fname, const char *target, FILE *f);

bfd *
bfd_openr(const char *fname, const char *target);

void
bfd_perror(const char *msg);

const char *
bfd_printable_name(bfd *abfd);

const bfd_arch_info_type *
bfd_scan_arch(const char *str);

bfd_boolean
bfd_set_section_alignment(bfd *abfd, asection *sect, unsigned int val);

bfd_size_type
bfd_section_size(bfd *abfd, asection *sect);

bfd_boolean
bfd_set_section_flags(bfd *abfd, asection *sect, flagword flags);

bfd_boolean
bfd_set_section_size(bfd *abfd, asection *sect, bfd_size_type val);

bfd_boolean
bfd_set_section_vma(bfd *abfd, asection *sect, bfd_vma vma);

asection *
bfd_set_section(asymbol *sym, asection *sect);

#define xmalloc  malloc
#define xrealloc realloc

/*
 * The following types and routines are NOT in BFD -- they
 * are added for ease of portability
 */

/* The code where relocations are dealt with in bfdstuff is
 * very concentrated.
 * We can be more efficient and save memory if we don't emulate
 * the BFD API.
 * Hence we provide an alternative.
 */

/* Table with relocations */
typedef struct pmbfd_areltab pmbfd_areltab;

typedef union  pmbfd_arelent pmbfd_arelent;

/* generate reloc table */
long
pmbfd_canonicalize_reloc(bfd *abfd, asection *sec, pmbfd_areltab *tab, asymbol **syms);

/* iterator over reloc table */
pmbfd_arelent *
pmbfd_reloc_next(bfd *abfd, pmbfd_areltab *tab, pmbfd_arelent *prev);


bfd_reloc_status_type
pmbfd_perform_relocation(bfd *abfd, pmbfd_arelent *reloc, asymbol *sym, asection *input_section);

int
pmbfd_reloc_get_sym_idx(bfd *abfd, pmbfd_arelent *reloc);

bfd_size_type
pmbfd_reloc_get_address(bfd *abfd, pmbfd_arelent *reloc);

const char *
pmbfd_reloc_get_name(bfd *abfd, pmbfd_arelent *reloc);

symvalue
pmbfd_asymbol_set_value(asymbol *sym, symvalue v);

/* experimental stuff to keep libopcodes happy */
#define BFD_DEFAULT_TARGET_SIZE  32

#define CONST_STRNEQ(s1,s2) (0 == strncmp((s1),(s2),sizeof(s2)-1))
static inline int
sprintf_vma(char *s, bfd_vma vma)
{
	return sprintf(s,"0x%08lx",vma);
}

/* Byte swapping */

/* Read big endian into host order    */
bfd_vma bfd_getb32(const void *p);
/* Read little endian into host order */
bfd_vma bfd_getl32(const void *p);

#define bfd_octets_per_byte(abfd) (1)


/* CPU-specific instruction sets      */
unsigned bfd_m68k_mach_to_features (int);

/* Not in BFD; implemented so we can create a 'objdump'-compatible printout */
file_ptr
pmbfd_get_section_filepos(bfd *abfd, asection *section);

/* Not in BFD; create object attributes tables (see pmelf.h)
 * from a '.gnu.attributes' section.
 *
 * RETURNS: attribute set or NULL (if no '.gnu.attributes'
 *          section is present or if it could not be parsed.
 *
 * NOTE:    the caller is responsible for releasing resources
 *          by calling pmelf_destroy_attribute_set() when the
 *          table is no longer used.
 *
 *          The return value is really a (Pmelf_attribute_set *).
 */
void *
pmbfd_get_file_attributes(bfd *abfd);

#ifdef __cplusplus
};
#endif

#endif
